<launch>

  <arg name="use_map_topic"   default="false"/>
  <arg name="scan_topic"      default="/kobuki/laser/scan"/>
  <arg name="initial_pose_x"  default="0.0"/>
  <arg name="initial_pose_y"  default="-1.0"/>
  <arg name="initial_pose_a"  default="0.0"/>
  <!-- for total uncertainty: in a state space 10 mt use cov = 10000 -->
  <arg name="initial_cov_xx"  default="10"/>
  <arg name="initial_cov_yy"  default="10"/>
  <arg name="initial_cov_aa"  default="0.068538919"/>
  <arg name="min_particles"   default="600"/>
  <arg name="max_particles"   default="2000"/>
  <arg name="resample_interval" default="6"/>

  <arg name="odom_frame_id"   default="/odom"/>
  <arg name="base_frame_id"   default="/base_footprint"/>
  <arg name="global_frame_id" default="map"/>

  <node pkg="amcl" type="amcl" name="robot_amcl">
    <param name="use_map_topic"             value="$(arg use_map_topic)"/>
    <!-- Publish scans from best pose at a max of 10 Hz -->
    <param name="odom_model_type"           value="diff-corrected"/>
    <param name="odom_alpha5"               value="0.1"/>
    <param name="gui_publish_rate"          value="10.0"/>
    <param name="laser_max_beams"             value="60"/>
    <param name="laser_max_range"           value="12.0"/>
    <param name="min_particles"             value="$(arg min_particles)"/>
    <param name="max_particles"             value="$(arg max_particles)"/>
    <param name="kld_err"                   value="0.05"/>
    <param name="kld_z"                     value="0.99"/>
    <param name="odom_alpha1"               value="0.05"/>
    <param name="odom_alpha2"               value="0.05"/>
    <!-- translation std dev, m -->
    <param name="odom_alpha3"               value="0.10"/>
    <param name="odom_alpha4"               value="0.05"/>
    <param name="laser_z_hit"               value="0.5"/>
    <param name="laser_z_short"             value="0.05"/>
    <param name="laser_z_max"               value="0.05"/>
    <param name="laser_z_rand"              value="0.5"/>
    <param name="laser_sigma_hit"           value="0.2"/>
    <param name="laser_lambda_short"        value="0.1"/>
    <param name="laser_model_type"          value="likelihood_field"/>
    <!-- <param name="laser_model_type" value="beam"/> -->
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="update_min_d"              value="0.25"/>
    <param name="update_min_a"              value="0.2"/>
    <param name="odom_frame_id"             value="$(arg odom_frame_id)"/>
    <param name="base_frame_id"             value="$(arg base_frame_id)"/>
    <param name="global_frame_id"           value="$(arg global_frame_id)"/>
    <!-- larger resample_interval values lead to slower particle convergence -->
    <param name="resample_interval"         value="$(arg resample_interval)"/>
    <!-- Increase tolerance because the computer can get quite busy -->
    <param name="transform_tolerance"       value="1.0"/>
    <param name="recovery_alpha_slow"       value="0.0"/>
    <param name="recovery_alpha_fast"       value="0.0"/>
    <param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
    <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
    <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>
    <param name="initial_cov_xx"            value="$(arg initial_cov_xx)"/>
    <param name="initial_cov_yy"            value="$(arg initial_cov_yy)"/>
    <param name="initial_cov_aa"            value="$(arg initial_cov_aa)"/>
    <remap from="scan"                      to="$(arg scan_topic)"/>
    <remap from="initialpose"               to="/initialpose"/>
    <remap from="amcl_pose"                 to="/amcl_pose"/>
    <remap from="particlecloud"             to="/particlecloud"/>

  </node>

</launch>